### DISCUSSION TITLE
针对 mpv 弹幕播放需求的脚本推荐

### DISCUSSION THREAD

**Discussion by dyphire** [dyphire/mpv-config#105] [General]
## bilibili 和动画疯的在线视频的弹幕播放脚本

https://github.com/itKelis/MPV-Play-BiliBili-Comments
https://github.com/s594569321/MPV-Play-BAHA-Comments

## 自动加载本地播放目录中同名的 xml 弹幕文件的弹幕脚本

https://github.com/dyphire/mpv-load-danmaku

## 为本地番剧自动加载 [dandanplay](https://api.dandanplay.net/swagger/ui/index) 弹幕的弹幕插件脚本

基于文件名和文件哈希值匹配弹幕：https://github.com/rkscv/danmaku
- 为 emby ([tsukimi](https://github.com/tsukinaha/tsukimi)) 实现适配的修改版：https://github.com/Kosette/danmaku

## 为番剧自动加载 [dandanplay](https://api.dandanplay.net/swagger/ui/index) 弹幕的 uosc 弹幕插件脚本

https://github.com/Tony15246/uosc_danmaku
- 自动/手动加载时的文件哈希值匹配
- 哈希匹配无结果时解析媒体名自动匹配
- 通过 uosc 弹幕菜单手动搜索关联弹幕
- 手动搜索时的自动解析预填充番剧名称（本地及 URL）
 - 手动搜索支持电影电视剧的弹幕搜索
- 记忆和继承同目录的弹幕关联
- url 视频文件的弹幕关联记忆及继承（已测试场景：emby, `magnet:?`）
- 自动/手动加载本地播放目录中同名的 xml 弹幕文件
- 支持 bilibili 和动画疯的在线视频的弹幕加载
- 自定义弹幕 API 服务地址（可支持非番剧的弹幕关联加载）
- uosc 不可用时尝试调用 mpv 内部的`mp.input`来渲染菜单

**Comment by zhongfly** [comment]
正好前几天看到https://github.com/itKelis/MPV-Play-BiliBili-Comments 就有一个想法：能不能借助yt-dlp来下载弹幕文件到本地，然后调用转换工具转为ass…这样就能做成一个适用多个网站的脚本
不过由于我基本不用mpv看在线视频，所以也不愿意额外花时间去做这个(好像已经有类似的实现了

另外说是win10某个版本起就自带curl，也可以用来下载文件。或者是直接用powershell来下载

**Reply by dyphire** [reply]
直接下载转成 ass 保存不太推荐，使用我推荐的第二个脚本要更优雅些

ps：我也基本不用 mpv 看在线视频

**Reply by zhongfly** [reply]
也没啥不优雅的，最后不都是要在本地新建一个ass文件吗，从结果来说都优雅不了
每个网站都单独一个脚本我觉得也不怎么优雅

从效果看，还是直接在线看好，还能保存/同步进度

**Reply by dyphire** [reply]
第二个脚本不是新建一个 ass 文件，它不是通过加载 ass 字幕轨道渲染的，而是通过 osd-overlay 的 ass-event 方式渲染的，比较巧妙。

这种方式有几个好处：平滑度不像 ass 字幕一样完全受制于视频帧率，OSD 由 video-sync 控制；渲染不局限于视频画面内；缩放不受视频分辨率影响只匹配 OSD 大小；不占用字幕轨道避免可能的脚本兼容性问题。但也存在缺点：这种方式似乎渲染不了静态的弹幕，都当作滚动弹幕处理了；无法处理弹幕延迟

**Reply by dyphire** [reply]
> 每个网站都单独一个脚本我觉得也不怎么优雅

yt-dlp 不支持动画疯，所以使用 yt-dlp 下载弹幕转成 ass 字幕保存的方式也无法适用各个在线播放场景。不过反正基本不用 mpv 看在线视频，所以不是很在意在线播放的弹幕方案

**Reply by ghost** [reply]
> 这种方式似乎渲染不了静态的弹幕，都当作滚动弹幕处理了；无法处理弹幕延迟

其实可以渲染静态的弹幕，只不过为了防挡字幕，我统一当作滚动弹幕处理了。由于弹幕的像素长度是借助 [unicode-segmentation](https://github.com/unicode-rs/unicode-segmentation) 这个库做近似的估计的，静态弹幕在水平方向上可能无法严格地居中。

弹幕延迟指的是什么？

**Reply by dyphire** [reply]
> 弹幕延迟指的是什么？

弹弹play 匹配的弹幕的时间戳可能和本地视频的不一致，比如 b 站的删减源的弹幕亦或者 b 站片头有多几秒到十几秒的内容就会导致时间戳匹配不上，弹幕内容和本地视频的画面不匹配。

通过字幕的加载方式可以通过调节字幕延迟来修正时间戳，osd-overlay 的时间戳似乎是脚本自行控制的，应该可以通过提供 key-binding 来调整事件的时间戳

**Reply by ghost** [reply]
弹弹 play 的视频匹配 API 的返回值里有一个 `shift` 参数，可能是给弹幕的时间戳做全局位移用的，但我没在文档里找到具体的说明，也没遇到过需要位移的情况，所以就没用它。至于视频中间的删减，API 的返回值里没有相关的参数，如果服务端没有处理的话应该就没办法了吧。

**Reply by dyphire** [reply]
字幕的话可以通过 sub-delay 或 secondary-sub-delay 的命令手动调整延迟来同步，osd-overlay 估计只能自己实现了

> 弹弹 play 的视频匹配 API 的返回值里有一个 `shift` 参数，可能是给弹幕的时间戳做全局位移用的，但我没在文档里找到具体的说明，也没遇到过需要位移的情况，所以就没用它

或许通过提供  key-binding 来支持手动调整 `shift` 就够了，效果应该和字幕的 sub-delay 命令差不多

**Reply by ghost** [reply]
> 静态弹幕在水平方向上可能无法严格地居中

基于 `\an8` （顶部中间对齐）计算 `\pos` 标签的参数就没有这个问题了。
